csharp/src/Ice/Communicator.cs

```diff
@@ -147,7 +147,17 @@ private class ObserverUpdater : Instrumentation.IObserverUpdater
         public ToStringMode ToStringMode { get; }
 
         // The communicator's cancellation token is notified of cancellation when the communicator is destroyed.
-        internal CancellationToken CancellationToken => _cancellationTokenSource.Token;
+        internal CancellationToken CancellationToken
+        {
+            get
+            {
+                if (_disposeTask != null)
+                {
+                    throw new CommunicatorDisposedException();
+                }
```

Comment:
Why is this needed? Won't the cancellation token be canceled anyway if the communicator is disposed? There's no synchronization on the _mutex there so this doesn't appear to be very useful.

Reply:
The cancelation token will be disposed and will throw ObjectDisposedException, when trying to make an invocation with a proxy, maybe we can check this when creating the linked token

Reply:
We could fix the communicator to assign the token on construction or add try/catch handler to rethrow CommunicatorDisposedException to prevent this. Checking for _disposeTask is not fool proof since you might still get ObjectDisposedException.